src/OpenSage.Game/Graphics/Cameras/RtsCameraController.cs

```diff
@@ -160,31 +167,66 @@ private static float GetKeyMovement(in CameraInputState inputState, Key positive
             return 0;
         }
 
-        void ICameraController.UpdateCamera(Camera camera, in CameraInputState inputState, in TimeInterval gameTime)
+        void ICameraController.UpdateCameraInput(Camera camera, in CameraInputState inputState, in TimeInterval gameTime)
         {
             if (inputState.LeftMouseDown && inputState.PressedKeys.Contains(Key.AltLeft) || inputState.MiddleMouseDown)
             {
                 RotateCamera(inputState.DeltaX, inputState.DeltaY);
+                _panDirection = CameraPanDirection.None;
             }
             else
             {
                 // tested in Zero Hour - rotation always takes precedence, and all panning is halted when rotating.
-                float forwards, right;
+                float up, right;
                 if (inputState.RightMouseDown)
                 {
-                    forwards = -inputState.DeltaY;
-                    right = inputState.DeltaX;
+                    up = -MousePanSpeed * inputState.DeltaY;
+                    right = MousePanSpeed * inputState.DeltaX;
+                }
+                else
+                {
+                    OpenSage.Mathematics.Rectangle bounds = _panel.Frame;
+
+                    if (inputState.LastX < PanBorderWith) right = -1;
+                    else if (inputState.LastX > bounds.Width - PanBorderWith) right = 1;
+                    else right = GetKeyMovement(inputState, Key.Right, Key.Left);
+
+                    if (inputState.LastY < PanBorderWith) up = 1;
+                    else if (inputState.LastY > bounds.Height - PanBorderWith) up = -1;
+                    else up = GetKeyMovement(inputState, Key.Up, Key.Down);
+                }
+                if (up != 0 || right != 0)
+                {
+                    const float thresh = 0.5f;
+                    if (up > thresh)
+                    {
+                        if (right > thresh) _panDirection = CameraPanDirection.RightUp;
+                        else if (right < -thresh) _panDirection = CameraPanDirection.LeftUp;
+                        else _panDirection = CameraPanDirection.Up;
+                    }
+                    else if (up < -thresh)
+                    {
+                        if (right > thresh) _panDirection = CameraPanDirection.RightDown;
+                        else if (right < -thresh) _panDirection = CameraPanDirection.LeftDown;
+                        else _panDirection = CameraPanDirection.Down;
+                    }
+                    else
+                    {
+                        if (right >= 0) _panDirection = CameraPanDirection.Right;
```

Comment:
there is no threshold for right & left only for up & down?

Reply:
That's original game behavior. Scroll cursor always visible when right mouse pressed (by default is right arrow). So threshold  needs only to distinguish vertical movement from horizontal.